File: elixir/house/lib/house.ex
defmodule House do
  @objects_and_action_suffered [
    {"house that Jack built.", "lay in"},
    {"malt", "ate"},
    {"rat", "killed"},
    {"cat", "worried"},
    {"dog", "tossed"},
    {"cow with the crumpled horn", "milked"},
    {"maiden all forlorn", "kissed"},
    {"man all tattered and torn", "married"},
    {"priest all shaven and shorn", "woke"},
    {"rooster that crowed in the morn", "kept"},
    {"farmer sowing his corn", "belonged to"},
    {"horse and the hound and the horn", ""}
  ]

  @doc """
  Return verses of the nursery rhyme 'This is the House that Jack Built'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    "#{start..stop |> Enum.map(&generate_verse/1) |> Enum.join("\n")}\n"
  end

  @spec generate_verse(non_neg_integer()) :: String.t()
  defp generate_verse(verse_number) do
    1..verse_number
    |> Enum.map(fn line -> generate_line(line, verse_number) end)
    |> Enum.join(" ")
  end

  @spec generate_line(non_neg_integer(), non_neg_integer()) :: String.t()
  defp generate_line(line_number, verse_number)

  defp generate_line(1, verse_number),
    do: (verse_number - 1) |> get_object() |> this_is()

  defp generate_line(line_number, verse_number) do
    {object, action} =
      @objects_and_action_suffered
      |> Enum.at(verse_number - line_number)

    that_do_to(action, object)
  end

  @spec this_is(String.t()) :: String.t()
  defp this_is(object),
    do: "This is the #{object}"

  @spec that_do_to(Strint.t(), Strint.t()) :: String.t()
  defp that_do_to(action, subject),
    do: "that #{action} the #{subject}"

  @spec get_object(String.t()) :: String.t()
  defp get_object(line) do
    {object, _} = @objects_and_action_suffered |> Enum.at(line)
    object
  end
end

File: lua/house/house.lua
--- A module for generating the nursery rhyme "This Is the House That Jack Built".
local house = {}

local that_do, this_is

--- A constant table containing the rhyming pairs for the song.
--- Each pair consists of an action and a subject.
local PAIRS <const> = {
    { "lay in",      "house that Jack built" },
    { "ate",         "malt" },
    { "killed",      "rat" },
    { "worried",     "cat" },
    { "tossed",      "dog" },
    { "milked",      "cow with the crumpled horn" },
    { "kissed",      "maiden all forlorn" },
    { "married",     "man all tattered and torn" },
    { "woke",        "priest all shaven and shorn" },
    { "kept",        "rooster that crowed in the morn" },
    { "belonged to", "farmer sowing his corn" },
    { "",            "horse and the hound and the horn" }
}

--- Generates a specific verse of the rhyme.
--- @param which integer The index of the verse to generate.
--- @return string -- The formatted verse.
house.verse = function(which)
    local lines = {}

    table.insert(lines, this_is(PAIRS[which][2], which == 1))

    for i = which - 1, 1, -1 do
        table.insert(lines, that_do(PAIRS[i][1], PAIRS[i][2], i == 1))
    end

    return table.concat(lines, "\n")
end

--- Generates the complete nursery rhyme, from start to finish.
--- @return string -- The full, formatted rhyme.
house.recite = function()
    local verses = {}

    for i = 1, #PAIRS, 1 do
        table.insert(verses, house.verse(i))
    end

    return table.concat(verses, "\n")
end

--- A constant format string for the lines of the form "that [action] the [subject]".
local THAT_LINE_FORMAT <const> = "that %s the %s%s"
--- A constant format string for the opening line of a verse: "This is the [thing]".
local THIS_LINE_FORMAT <const> = "This is the %s%s"

--- Formats a single line of the form "that [action] the [subject]".
--- @param action string The action verb.
--- @param subject string The subject of the line.
--- @param terminal boolean True if the line should end with a period.
--- @return string -- The formatted line.
function that_do(action, subject, terminal)
    local terminator = terminal and "." or ""

    return THAT_LINE_FORMAT:format(action, subject, terminator)
end

--- Formats the opening line of a verse.
--- @param thing string The item being introduced.
--- @param terminal boolean True if the line should end with a period.
--- @return string -- The formatted line.
function this_is(thing, terminal)
    local terminator = terminal and "." or ""

    return THIS_LINE_FORMAT:format(thing, terminator)
end

return house
